File: client/python/mesh/prsim/cache.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#
import time
from abc import abstractmethod, ABC
from typing import Any, Type

from mesh.kinds import Entity
from mesh.kinds.entity import CacheEntity
from mesh.macro import spi, mpi, T


@spi(name="mesh")
class Cache(ABC):

    @abstractmethod
    @mpi("mesh.cache.get")
    def get(self, key: str) -> CacheEntity:
        """
        :param key:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.cache.put")
    def put(self, cell: CacheEntity) -> None:
        """
        :param cell:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.cache.remove")
    def remove(self, key: str):
        """
        :param key:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.cache.incr")
    def incr(self, key: str, value: int, duration: int) -> int:
        """
        :param key:
        :param value:
        :param duration:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.cache.decr")
    def decr(self, key: str, value: int, duration: int) -> int:
        """
        :param key:
        :param value:
        :param duration:
        :return:
        """
        pass

    def get_with_type(self, key: str, kind: Type[T]) -> T:
        """
        Get cache value with generic type.
        :param key:
        :param kind:
        :return:
        """
        ce = self.get(key)
        if ce is None:
            return None
        return ce.entity().try_read_object(kind)

    def put_with_duration(self, key: str, value: Any, duration: int):
        """
        Put value to cache with expire duration.
        :param key:
        :param value:
        :param duration:
        :return:
        """
        cell = CacheEntity()
        cell.version = "1.0.0"
        cell.entity = Entity.wrap(value)
        cell.timestamp = int(time.time() * 1000)
        cell.duration = duration
        cell.key = key
        self.put(cell)

    def compute_if_absent(self, key: str, kind: Type[T], duration: int, fn):
        """
        Default compute and put if absent.
        :param key:
        :param kind:
        :param duration:
        :param fn:
        :return:
        """
        value = self.get_with_type(key, kind)
        if value is not None:
            return value
        value = fn(key)
        if value is not None:
            self.put_with_duration(key, value, duration)
        return value

File: client/java/src/main/java/com/be/mesh/client/prsim/Cache.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.prsim;

import com.be.mesh.client.annotate.Index;
import com.be.mesh.client.annotate.MPI;
import com.be.mesh.client.annotate.SPI;
import com.be.mesh.client.mpc.Types;
import com.be.mesh.client.struct.CacheEntity;
import com.be.mesh.client.struct.Entity;

import java.time.Duration;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;

/**
 * @author coyzeng@gmail.com
 */
@SPI("mesh")
public interface Cache {

    @MPI("mesh.cache.get")
    CacheEntity get(@Index(0) String key);

    @MPI("mesh.cache.put")
    void put(@Index(0) CacheEntity cell);

    @MPI("mesh.cache.remove")
    void remove(@Index(0) String key);

    @MPI("mesh.cache.incr")
    long incr(@Index(0) String key, @Index(1) long value);

    @MPI("mesh.cache.decr")
    long decr(@Index(0) String key, @Index(1) long value);

    /**
     * Keys the cache key set.
     */
    @MPI("mesh.cache.keys")
    List<String> keys(@Index(0) String pattern);

    /**
     * HGet get value in hash
     */
    @MPI("mesh.cache.hget")
    CacheEntity hget(String key, String name);

    /**
     * HSet put value in hash
     */
    @MPI("mesh.cache.hset")
    void hset(String key, CacheEntity cell);

    /**
     * HDel put value in hash
     */
    @MPI("mesh.cache.hdel")
    void hdel(String key, String name);

    /**
     * HKeys get the hash keys
     */
    @MPI("mesh.cache.hkeys")
    List<String> hkeys(String key);

    /**
     * Get by codec.
     */
    default <T> T get(String key, Types<T> type) {
        return Optional.ofNullable(this.get(key)).map(entity -> entity.getEntity().tryReadObject(type)).orElse(null);
    }

    /**
     * Put by codec.
     */
    default void put(String key, Object value, Duration duration) {
        CacheEntity cell = new CacheEntity();
        cell.setVersion("1.0.0");
        cell.setEntity(Entity.wrap(value));
        cell.setTimestamp(System.currentTimeMillis());
        cell.setDuration(duration.toMillis());
        cell.setKey(key);
        this.put(cell);
    }

    /**
     * Default compute and put if absent.
     */
    default <T> T computeIfAbsent(String key, Types<T> type, Duration duration, Function<String, T> fn) {
        T value = get(key, type);
        if (null != value) {
            return value;
        }
        value = fn.apply(key);
        if (null != value) {
            put(key, value, duration);
        }
        return value;
    }
}
